Create config and infrastructure for a What's new in docs for a release (#30545)
dotnet/docs: docs/whats-new/dotnet-7-docs.md

title: ".NET docs: What's new for .NET 7"
description: "What's new in the .NET docs for the .NET 7 release."
ms.date: 11/08/2022
---

# .NET docs: What's new for the .NET 7 release

Welcome to what's new in the .NET docs for the .NET 7 release. This article lists some of the major changes to docs related to .NET feature updates.

## .NET breaking changes

- [Breaking changes in .NET 7](../core/compatibility/7.0.md)

## .NET fundamentals

### New articles

- [What's new in .NET 7](../core/whats-new/dotnet-7.md)
- [Introduction to AOT warnings](../core/deploying/native-aot/fixing-warnings.md)
- [Diagnostic monitoring and collection utility (dotnet-monitor)](../core/diagnostics/dotnet-monitor.md)
- [Containerize a .NET app with dotnet publish](../core/docker/publish-as-container.md)
- [System.Threading.Channels library](../core/extensions/channels.md)
- [Rate limit an HTTP handler in .NET](../core/extensions/http-ratelimiter.md)
- [Upgrade a WCF Server-side Project to use CoreWCF on .NET 6](../core/porting/upgrade-assistant-wcf.md)
- [dotnet workload command](../core/tools/dotnet-workload.md)
- [NuGet signed package verification](../core/tools/nuget-signed-package-verification.md)

> - [.NET SDK error list](../core/tools/sdk-errors/index.md)

Reviewer: Not a .NET 7 release item
```suggestion
```

- [dotnet new list](../core/tools/dotnet-new-list.md) - New syntax for `dotnet new`
- [dotnet new search](../core/tools/dotnet-new-search.md) - New syntax for `dotnet new`
- [dotnet new uninstall](../core/tools/dotnet-new-uninstall.md) - New syntax for `dotnet new`
- [dotnet new update](../core/tools/dotnet-new-update.md) - New syntax for `dotnet new`
- [dotnet new &lt;TEMPLATE&gt;](../core/tools/dotnet-new.md) - New syntax for `dotnet new`
- [dotnet restore](../core/tools/dotnet-restore.md) - Document .NET 7 SDK changes to NuGet signed package verification
- [dotnet publish](../core/tools/dotnet-publish.md) - Document Publish UseCurrentRuntime
- [dotnet build](../core/tools/dotnet-build.md) - Document UseCurrentRuntime
- [dotnet watch](../core/tools/dotnet-watch.md) - Note that --non-interactive is .NET 7
- [dotnet nuget sign](../core/tools/dotnet-nuget-sign.md) - Document .NET 7 SDK changes to NuGet signed package verification
- [dotnet nuget trust](../core/tools/dotnet-nuget-trust.md) - Document .NET 7 SDK changes to NuGet signed package verification
- [dotnet nuget verify](../core/tools/dotnet-nuget-verify.md) - Document .NET 7 SDK changes to NuGet signed package verification

> - [.NET SDK error list](../core/tools/sdk-errors/index.md)
>   - Add new errors/warnings to NETSDK error index
>   - Document selected NETSDK error messages
>   - List of NETSDK error messages

Reviewer: ```suggestion
- [.NET SDK error list](../core/tools/sdk-errors/index.md) - Add new errors/warnings to NETSDK error index
```

- [Parsing numeric strings in .NET](../standard/base-types/parsing-numeric.md) - Generic math docs
- [Generics in .NET](../standard/generics/index.md) - Generic math docs
- [Generic interfaces in .NET](../standard/generics/interfaces.md) - Generic math docs
- [Runtime libraries overview](../standard/runtime-libraries-overview.md) - New `HttpClient` article
- [Target frameworks in SDK-style projects](../standard/frameworks.md) - Update TFMs for .NET 7
- [Source generation for platform invokes](../standard/native-interop/pinvoke-source-generation.md)
  - Add SYSLIB diagnostics for .NET 7 source generators
  - Custom marshaller tutorial
  - Basic overview of p/invoke source generation
- [Tutorial: Use custom marshallers in source-generated P/Invokes](../standard/native-interop/tutorial-custom-marshaller.md) - Custom marshaller tutorial
- [How to choose reflection or source generation in System.Text.Json](../standard/serialization/system-text-json/source-generation-modes.md) - Source generation for STJ update
- [How to use source generation in System.Text.Json](../standard/serialization/system-text-json/source-generation.md) - Source generation for STJ update

> - [How to write custom converters for JSON serialization (marshalling) in .NET](../standard/serialization/system-text-json/converters-how-to.md)
>   - Add section on Utf8JsonReader.CopyString
>   - Add contract customization article

Reviewer: ```suggestion
- [How to write custom converters for JSON serialization (marshalling) in .NET](../standard/serialization/system-text-json/converters-how-to.md) - Add section on Utf8JsonReader.CopyString
```

- [Single-file deployment and executable](../core/deploying/single-file/overview.md) - Update single file incompatibility docs
- [Prepare .NET libraries for trimming](../core/deploying/trimming/prepare-libraries-for-trimming.md)- Update trimming libraries docs
- [Trimming options](../core/deploying/trimming/trimming-options.md) - Cleanup trimming options and document TrimMode=full,partial
- [Tutorial: Containerize a .NET app](../core/docker/build-container.md) - Add `dotnet publish` article for .NET 7 containerization support.
- [Rate limit an HTTP handler in .NET](../core/extensions/http-ratelimiter.md) - New `RateLimiter` article showing a custom HTTP handler
- [Configuration providers in .NET](../core/extensions/configuration-providers.md) - Environment variable prefix breaking change
- [Use HTTP/3 with HttpClient](../core/extensions/httpclient-http3.md) - Update HTTP/3 support for HttpClient in .NET 7
- [IHttpClientFactory with .NET](../core/extensions/httpclient-factory.md) - New `HttpClient` article
- [Overview of .NET, MSBuild, and Visual Studio versioning](../core/porting/versioning-sdk-msbuild-vs.md)
  - Add preview SDK table
  - Formalize versioning support policy

> - [Overview of the .NET Upgrade Assistant](../core/porting/upgrade-assistant-overview.md)

Author: removing per comment:

```suggestion
```

> - [Overview of the .NET Upgrade Assistant](../core/porting/upgrade-assistant-overview.md) - Merge .NET 7 release branch into main

Reviewer 2: ```suggestion
- [Overview of the .NET Upgrade Assistant](../core/porting/upgrade-assistant-overview.md)
```

- [Upgrade a WCF Server-side Project to use CoreWCF on .NET 6](../core/porting/upgrade-assistant-wcf.md) - Add documentation for CoreWCF extension on Upgrade Assistant
- [.NET RID Catalog](../core/rid-catalog.md) - Add iOS and Android RIDs
- [Runtime configuration options for threading](../core/runtime-config/threading.md) - Document a new config switch added to the portable thread pool
- [MSBuild reference for .NET SDK projects](../core/project-sdk/msbuild-props.md)
  - Add three new properties
  - Add docs for project properties that control PDBs and XML publishing
  - Update the documentation for PublishRelease and PackRelease
  - Add DisableTransitiveProjectReferences
  - Cleanup trimming options and document TrimMode=full,partial
  - Document new PublishRelease and PackRelease properties

> - [How to enable tab completion for the .NET CLI](../core/tools/enable-tab-autocomplete.md)

Author: Removing per comment:

```suggestion
```

> - [How to enable tab completion for the .NET CLI](../core/tools/enable-tab-autocomplete.md) - Merge .NET 7 release branch into main

Reviewer 2: ```suggestion
- [How to enable tab completion for the .NET CLI](../core/tools/enable-tab-autocomplete.md)
```

- [Diagnostics client library](../core/diagnostics/diagnostics-client-library.md) - Add dotnet-monitor .NET tool documentation
- [Collect diagnostics in containers](../core/diagnostics/diagnostics-in-containers.md) - Add dotnet-monitor .NET tool documentation
- [What diagnostic tools are available in .NET Core?](../core/diagnostics/index.md) - Add dotnet-monitor .NET tool documentation
- [Source-generator diagnostics in .NET 6+](../fundamentals/syslib-diagnostics/source-generator-overview.md) - Add SYSLIB diagnostics for .NET 7 source generators
- [Obsolete features in .NET 5+](../fundamentals/syslib-diagnostics/obsoletions-overview.md)
  - Add two new SYSLIB warnings
  - XmlSecureResolver breaking change
  - Add SYSLIB0041
  - Add SYSLIB0040
  - Add syslib0038 and syslib0039
- [SYSLIB0012: Assembly.CodeBase and Assembly.EscapedCodeBase are obsolete](../fundamentals/syslib-diagnostics/syslib0012.md) - Add two new SYSLIB warnings
- [SYSLIB0043: ECDiffieHellmanPublicKey.ToByteArray is obsolete](../fundamentals/syslib-diagnostics/syslib0043.md)
  - XmlSecureResolver breaking change
  - Add SYSLIB0041

> - [Overview of .NET source code analysis](../fundamentals/code-analysis/overview.md) - Update list of rules enabled for .NET 7

Author: Removing per comment:

```suggestion
```

> - [Overview of .NET source code analysis](../fundamentals/code-analysis/overview.md) - Merge .NET 7 release branch into main

Reviewer 2: ```suggestion
- [Overview of .NET source code analysis](../fundamentals/code-analysis/overview.md) - Update list of rules enabled for .NET 7
```

- [Built-in numeric conversions (C# reference)](../csharp/language-reference/builtin-types/numeric-conversions.md) - Add updates for numeric IntPtr
- [Arithmetic operators (C# reference)](../csharp/language-reference/operators/arithmetic-operators.md) - Operator updates to support generic math
- [Determine caller information using attributes interpreted by the C# compiler](../csharp/language-reference/attributes/caller-information.md) - Extended nameof parameter scope
- [Attributes for null-state static analysis interpreted by the C# compiler](../csharp/language-reference/attributes/nullable-analysis.md) - Extended nameof parameter scope
- [delegate operator (C# reference)](../csharp/language-reference/operators/delegate-operator.md) - Document new method group conversion
- [Generics and Attributes (C# Programming Guide)](../csharp/programming-guide/generics/generics-and-attributes.md) - generic attributes are allowed in C# 11

## Microsoft Orleans

### New articles

- [What's new in Orleans 7.0](../orleans/whats-new-in-orleans.md)

### Updated articles

- [Orleans transactions](../orleans/grains/transactions.md) - Merge .NET 7 release branch into main

> - [Microsoft Orleans documentation](../orleans/index.yml) - What is new Orleans

Reviewer 2: ```suggestion
- [Microsoft Orleans documentation](../orleans/index.yml) - What's new Orleans
```

- [SYSLIB diagnostics for p/invoke source generation](../fundamentals/syslib-diagnostics/syslib1050-1069.md)
- [SYSLIB diagnostics for JavaScript interop source generation](../fundamentals/syslib-diagnostics/syslib1070-1089.md)
- [SYSLIB0038: SerializationFormat.Binary is obsolete](../fundamentals/syslib-diagnostics/syslib0038.md)
- [SYSLIB0039: SslProtocols.Tls and SslProtocols.Tls11 are obsolete](../fundamentals/syslib-diagnostics/syslib0039.md)
- [SYSLIB0040: EncryptionPolicy.NoEncryption and EncryptionPolicy.AllowNoEncryption are obsolete](../fundamentals/syslib-diagnostics/syslib0040.md)
- [SYSLIB0041: Some Rfc2898DeriveBytes constructors are obsolete](../fundamentals/syslib-diagnostics/syslib0041.md)
- [SYSLIB0042: FromXmlString and ToXmlString on ECC types are obsolete](../fundamentals/syslib-diagnostics/syslib0042.md)
- [SYSLIB0043: ECDiffieHellmanPublicKey.ToByteArray is obsolete](../fundamentals/syslib-diagnostics/syslib0043.md)
- [SYSLIB0044: AssemblyName.CodeBase and AssemblyName.EscapedCodeBase are obsolete](../fundamentals/syslib-diagnostics/syslib0044.md)
- [SYSLIB0046: ControlledExecution.Run should not be used](../fundamentals/syslib-diagnostics/syslib0046.md)

> - [SYSLIB0047: XmlSecureResolver is obsolete](../fundamentals/syslib-diagnostics/syslib0047.md)

Author: Is there an overview for this section? Would that simplify the output?

Reviewer 2: There is an overview, but it includes more than just what's new for .NET 7, so I think this is correct as is.

- [CA2019: `ThreadStatic` fields should not use inline initialization](../fundamentals/code-analysis/quality-rules/ca2019.md)
- [CA2259: Ensure `ThreadStatic` is only used with static fields](../fundamentals/code-analysis/quality-rules/ca2259.md)
- [CA1420: Property, type, or attribute requires runtime marshalling](../fundamentals/code-analysis/quality-rules/ca1420.md)
- [CA1422: Validate platform compatibility - obsoleted APIs](../fundamentals/code-analysis/quality-rules/ca1422.md)
- [CA2260: Implement generic math interfaces correctly](../fundamentals/code-analysis/quality-rules/ca2260.md)
- [CA1854: Prefer the `IDictionary.TryGetValue(TKey, out TValue)` method](../fundamentals/code-analysis/quality-rules/ca1854.md)
- [CA1851: Possible multiple enumerations of `IEnumerable` collection](../fundamentals/code-analysis/quality-rules/ca1851.md)
- [Prefer 'null' check over type check (IDE0150)](../fundamentals/code-analysis/style-rules/ide0150.md)
- [Simplify property pattern (IDE0170)](../fundamentals/code-analysis/style-rules/ide0170.md)
- [Use tuple to swap values (IDE0180)](../fundamentals/code-analysis/style-rules/ide0180.md)

> - [Namespace declaration preferences (IDE0160 and IDE0161)](../fundamentals/code-analysis/style-rules/ide0160-ide0161.md)

Author: Same question: Is there an overview page we should use?

Reviewer 2: Do you mean just link to the overview? I don't think that would be ideal.

- [Single-file deployment and executable](../core/deploying/single-file/overview.md) - Update single file incompatibility docs
- [Prepare .NET libraries for trimming](../core/deploying/trimming/prepare-libraries-for-trimming.md)- Update trimming libraries docs
- [Trimming options](../core/deploying/trimming/trimming-options.md) - Cleanup trimming options and document TrimMode=full,partial
- [Tutorial: Containerize a .NET app](../core/docker/build-container.md) - Add `dotnet publish` article for .NET 7 containerization support.
- [Rate limit an HTTP handler in .NET](../core/extensions/http-ratelimiter.md) - New `RateLimiter` article showing a custom HTTP handler
- [Configuration providers in .NET](../core/extensions/configuration-providers.md) - Environment variable prefix breaking change
- [Use HTTP/3 with HttpClient](../core/extensions/httpclient-http3.md) - Update HTTP/3 support for HttpClient in .NET 7
- [IHttpClientFactory with .NET](../core/extensions/httpclient-factory.md) - New `HttpClient` article
- [Overview of .NET, MSBuild, and Visual Studio versioning](../core/porting/versioning-sdk-msbuild-vs.md)
  - Add preview SDK table
  - Formalize versioning support policy

> - [Overview of the .NET Upgrade Assistant](../core/porting/upgrade-assistant-overview.md) - Merge .NET 7 release branch into main

Author: Should this be included?  And should it have a different note about the change?

Reviewer 3: No, this should be removed. It wasn't updated for .NET 7.

- [Upgrade a WCF Server-side Project to use CoreWCF on .NET 6](../core/porting/upgrade-assistant-wcf.md) - Add documentation for CoreWCF extension on Upgrade Assistant
- [.NET RID Catalog](../core/rid-catalog.md) - Add iOS and Android RIDs
- [Runtime configuration options for threading](../core/runtime-config/threading.md) - Document a new config switch added to the portable thread pool
- [MSBuild reference for .NET SDK projects](../core/project-sdk/msbuild-props.md)
  - Add three new properties
  - Add docs for project properties that control PDBs and XML publishing
  - Update the documentation for PublishRelease and PackRelease
  - Add DisableTransitiveProjectReferences
  - Cleanup trimming options and document TrimMode=full,partial
  - Document new PublishRelease and PackRelease properties

> - [How to enable tab completion for the .NET CLI](../core/tools/enable-tab-autocomplete.md) - Merge .NET 7 release branch into main

Author: Should this be included?  And should it have a different note about the change?

Reviewer 3: Same here, this wasn't updated for .NET 7. It can be removed.

- [dotnet publish](../core/tools/dotnet-publish.md) - Document Publish UseCurrentRuntime
- [dotnet build](../core/tools/dotnet-build.md) - Document UseCurrentRuntime
- [dotnet watch](../core/tools/dotnet-watch.md) - Note that --non-interactive is .NET 7
- [dotnet nuget sign](../core/tools/dotnet-nuget-sign.md) - Document .NET 7 SDK changes to NuGet signed package verification
- [dotnet nuget trust](../core/tools/dotnet-nuget-trust.md) - Document .NET 7 SDK changes to NuGet signed package verification
- [dotnet nuget verify](../core/tools/dotnet-nuget-verify.md) - Document .NET 7 SDK changes to NuGet signed package verification
- [.NET SDK error list](../core/tools/sdk-errors/index.md)
  - Add new errors/warnings to NETSDK error index
  - Document selected NETSDK error messages
  - List of NETSDK error messages
- [NETSDK1100: Set the EnableWindowsTargeting property to true](../core/tools/sdk-errors/netsdk1100.md)
  - Add three new properties
  - Document selected NETSDK error messages
- [dotnet-coverage code coverage utility](../core/additional-tools/dotnet-coverage.md) - Adding new options and example scenarios

> - [.NET additional tools overview](../core/additional-tools/index.md) - Add dotnet-monitor .NET tool documentation

Author: Is there a reasonable overview page for the SDK updates?

Reviewer 2: @tdykstra

Reviewer 3: Both the _introduction.md_ and _sdk.md_ files were updated by Rich Lander, it wouldn't hurt to include them here. See https://github.com/dotnet/docs/commit/22aedbeaa493b18631686756174c30945993db3d

Reviewer: > Is there a reasonable overview page for the SDK updates?

I don't believe so, but [dotnet new &lt;TEMPLATE&gt;](../core/tools/dotnet-new.md#description) would work as an overview for all of the dotnet new articles.

- [Diagnostics client library](../core/diagnostics/diagnostics-client-library.md) - Add dotnet-monitor .NET tool documentation
- [Collect diagnostics in containers](../core/diagnostics/diagnostics-in-containers.md) - Add dotnet-monitor .NET tool documentation
- [What diagnostic tools are available in .NET Core?](../core/diagnostics/index.md) - Add dotnet-monitor .NET tool documentation
- [Source-generator diagnostics in .NET 6+](../fundamentals/syslib-diagnostics/source-generator-overview.md) - Add SYSLIB diagnostics for .NET 7 source generators
- [Obsolete features in .NET 5+](../fundamentals/syslib-diagnostics/obsoletions-overview.md)
  - Add two new SYSLIB warnings
  - XmlSecureResolver breaking change
  - Add SYSLIB0041
  - Add SYSLIB0040
  - Add syslib0038 and syslib0039
- [SYSLIB0012: Assembly.CodeBase and Assembly.EscapedCodeBase are obsolete](../fundamentals/syslib-diagnostics/syslib0012.md) - Add two new SYSLIB warnings
- [SYSLIB0043: ECDiffieHellmanPublicKey.ToByteArray is obsolete](../fundamentals/syslib-diagnostics/syslib0043.md)
  - XmlSecureResolver breaking change
  - Add SYSLIB0041

> - [Overview of .NET source code analysis](../fundamentals/code-analysis/overview.md) - Merge .NET 7 release branch into main

Author: Should this be included?  And should it have a different note about the change?

Reviewer 3: No, remove it.

- [Use local function instead of lambda (IDE0039)](../fundamentals/code-analysis/style-rules/ide0039.md) - Add missing code-style rules
- [Code-style rules](../fundamentals/code-analysis/style-rules/index.md)
  - Add missing code-style rules
  - Add IDE0160 and IDE0161.
- [Performance rules](../fundamentals/code-analysis/quality-rules/performance-warnings.md)
  - Add documentation for rule CA1854
  - Add document for CA1851
- [Language rules](../fundamentals/code-analysis/style-rules/language-rules.md)
  - Add missing code-style rules
  - Add IDE0160 and IDE0161.
- [Null-checking preferences](../fundamentals/code-analysis/style-rules/null-checking-preferences.md) - Add missing code-style rules
- [Pattern matching preferences](../fundamentals/code-analysis/style-rules/pattern-matching-preferences.md) - Add missing code-style rules
- [.NET runtime contention events](../fundamentals/diagnostics/runtime-contention-events.md) - Update the doc with the new version of ContentionStart
- [Error codes returned by package validation](../fundamentals/package-validation/diagnostic-ids.md)
  - add docs for diagnostics CP0012 and CP0013

>   - add docs for diagnostics CP0010 and CP0011

Author: A lot more diagnostic updates here. Is there an overview page that would be useful?

Reviewer 2: This is an overview page already.

- [Determine caller information using attributes interpreted by the C# compiler](../csharp/language-reference/attributes/caller-information.md) - Extended nameof parameter scope
- [Attributes for null-state static analysis interpreted by the C# compiler](../csharp/language-reference/attributes/nullable-analysis.md) - Extended nameof parameter scope
- [delegate operator (C# reference)](../csharp/language-reference/operators/delegate-operator.md) - Document new method group conversion
- [Generics and Attributes (C# Programming Guide)](../csharp/programming-guide/generics/generics-and-attributes.md) - generic attributes are allowed in C# 11

## F# language

### Updated articles

- [F# docs - get started, tutorials, reference.](../fsharp/index.yml) - Merge .NET 7 release branch into main
- [What's new in F# 6](../fsharp/whats-new/fsharp-6.md)
  - Merge .NET 7 release branch into main
  - Retire the "What's new in C# 8" article
- [What's new in F# 5](../fsharp/whats-new/fsharp-50.md) - Retire the "What's new in C# 8" article

## Microsoft Orleans

### New articles

- [What's new in Orleans 7.0](../orleans/whats-new-in-orleans.md)

### Updated articles

> - [Orleans transactions](../orleans/grains/transactions.md) - Merge .NET 7 release branch into main

Author: Should this be included?  And should it have a different note about the change?

Reviewer 2: ```suggestion
- [Orleans transactions](../orleans/grains/transactions.md) - Describe new APIs for .NET 7
```
